Huge merge

lalo opened the pull request:
This merge consists of:
- gbc c++ command line changes (adds --grpc)
- sync codegen
- basic thread_pool implementation + unit tests
- bond_utils.h for serialization and deserialization integration with grpc
- sync helloworld example that runs as a test

Commit:
Checkpoint work on gRPC output

Commit:
Checkpoint work on grpc for C++

Commit:
Add gRPC C++ example

Commit:
Save non-working C++ gRPC integration.

lalo committed:
[grpc] Add sync codegen

lalo committed:
[c++] Add thread pool

- add tests for thread pool

lalo committed:
Merge branch 'grpc_cpp' into grpc-merge

lalo committed:
[grpc c++] Add "sync" helloworld test

- Add ignore warning to codegen
- Replace dummy example with helloworld example
- Uses bond over grpc sync

lalo committed:
[grpc c++] Delete bogus example

- delete non functioning grpc ping pong

lalo committed:
[grpc c++] Remove unused .bond

chwarr commented on cpp/inc/bond/ext/grpc/bond_utils.h:
```diff
@@ -0,0 +1,86 @@
+  static Status Serialize(const bond::comm::message<T>& msg, grpc_byte_buffer** bp,
+                          bool* own_buffer) {
+    // figure out own buffer
+    *own_buffer = true;
```
Yes, this is correct. By setting this to true, we're telling gRPC to free the buffer by calling [`grpc_byte_buffer_destroy`](https://github.com/grpc/grpc/blob/ae95edb515c72a51655e4164f3e2a96e4e33677c/include/grpc%2B%2B/impl/codegen/call.h#L295-L310) when it's done sending.

chwarr commented on cpp/inc/bond/ext/grpc/bond_utils.h:
```diff
@@ -0,0 +1,86 @@
+
+    //todo: change cast to static
+    grpc_slice slice = grpc_slice_from_copied_buffer(
+        reinterpret_cast<const char*>(data.data()), data.length());
```
`data.content()` is what we want here, instead of casting `data.data()`.

chwarr commented on cpp/inc/bond/ext/grpc/bond_utils.h:
```diff
@@ -0,0 +1,86 @@
+
+    grpc_slice_unref(slice);
+
+    return Status(StatusCode::OK, "ok");
```
Minor: don't need the "ok" string here.

chwarr commented on cpp/inc/bond/ext/grpc/bond_utils.h:
```diff
@@ -0,0 +1,86 @@
+    bond::blob data = bond::blob(buff, GPR_SLICE_LENGTH(slice));
+    bond::CompactBinaryReader<bond::InputBuffer> cbreader(data);
+    bond::bonded<T> payload(cbreader);
+    //bond::Deserialize(cbreader, payload);
```
Can we delete this commented out code?

chwarr commented on cpp/inc/bond/ext/grpc/bond_utils.h:
```diff
@@ -0,0 +1,86 @@
+    boost::shared_ptr<char []> buff = boost::make_shared_noinit<char []>(GRPC_SLICE_LENGTH(slice));
+    std::memcpy(buff.get(), GRPC_SLICE_START_PTR(slice), GRPC_SLICE_LENGTH(slice));
+
+    // figure out ref counting for bytebuffer
```
Minor: prefix this with a TODO

chwarr commented on cpp/inc/bond/ext/grpc/bond_utils.h:
```diff
@@ -0,0 +1,86 @@
+
+    grpc_byte_buffer_destroy(buffer);
+
+    return Status(StatusCode::OK, "ok");
```
Minor: don't need "ok" string here either.

chwarr commented on cpp/inc/bond/ext/grpc/thread_pool.h:
```diff
@@ -0,0 +1,97 @@
+    // @remark:
+    //     Zero number of concurrent threads will allow simultaneous
+    //     execution of as many threads as many CPU/cores are available.
+    //     This is default value.
```
Minor: the "This is default value." comment just repeats what the code says pretty clearly. I'd remove it.

chwarr commented on cpp/inc/bond/ext/grpc/thread_pool.h:
```diff
@@ -0,0 +1,97 @@
+public:
+
+    //
+    // Constructs thread pool with specified number of threads
```
Minor: "Constructs _a_ thread pool with _the_ specified..."

chwarr commented on cpp/inc/bond/ext/grpc/thread_pool.h:
```diff
@@ -0,0 +1,97 @@
+    //
+    // Working threads.
+    //
+    std::list<boost::scoped_thread<boost::join_if_joinable>> m_threads;
```
The code doesn't appear to be making much use of the linked-list nature of `m_threads`. A `std::vector` looks fine here, unless performance testing shows otherwise. (We'll want to call `std::vector::reserve` before the `push_back` loop in the ctor, as we know exactly how many items we're going to push.)

chwarr replied:
Minor: `_threads` and `_work` for members variables to follow the convention in the core bond headers for private member variables.

chwarr commented on cpp/inc/bond/ext/grpc/thread_pool.h:
```diff
@@ -0,0 +1,97 @@
+        return *this;
+    }
+
+    //
```
These members looks like they can and should be private.

chwarr commented on cpp/test/grpc/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+    add_target_to_folder (${name})
+    add_dependencies (${name}
+        bond)
+        # grpc_test_codegen)
```
Should this comment be removed? The target_link_libraries below sets up a dependency for us.

chwarr commented on cpp/test/grpc/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+# Build common code into its own library.
+add_library (grpc_test_common EXCLUDE_FROM_ALL "main.cpp")
+add_target_to_folder (grpc_test_common)
+# add_dependencies(grpc_test_common grpc_test_codegen)
```
Remove this commented out code?

chwarr commented on cpp/test/grpc/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+target_link_libraries (grpc_test_common PRIVATE
+    ${Boost_UNIT_TEST_FRAMEWORK_LIBRARY})
+
+# add_bond_codegen (TARGET grpc_test_codegen grpc_test_common.bond GRPC)
```
Remove this commented out code?

chwarr commented on cpp/inc/bond/ext/grpc/thread_pool.h:
```diff
@@ -0,0 +1,97 @@
+#include <list>
+#include <thread>
+
+namespace bond { namespace ext
```
There's a mismatch between the namespace here and the file path. Intentional?

lalo replied:
sort of, didn't feel it should be on the grpc namespace but left it in that path since the path is not set in stone yet

chwarr replied:
Agreed that this isn't gRPC-specific. For now, let's just put it in bond/ext/thread_pool.h

chwarr commented on cpp/test/grpc/thread_pool.cpp:
```diff
@@ -0,0 +1,93 @@
+    void Initialize()
+    {
+        std::string name = "ThreadPool";
+        UnitTestSuite suite(name.c_str());
```
The `name` local doesn't appear to be needed. Can we just use `UnitTestSuite suite("gRPC ThreadPool");`?

chwarr commented on cpp/test/grpc/thread_pool.cpp:
```diff
@@ -0,0 +1,93 @@
+    void UseStdFunction()
+    {
+        bond::ext::thread_pool threads(1);
+        int sum = 0;
```
Can we be safe and use `std::atomic<int> sum` instead of doing this without any synchronization?

chwarr replied:
This has been changed to an `atomic<int>` in `FinishAllTasksAfterDelete` but not here. Intentional?

lalo replied:
yes, did not change it since its not being exercised

chwarr commented on cpp/test/grpc/thread_pool.cpp:
```diff
@@ -0,0 +1,93 @@
+
+        threads.schedule(std::bind(f_addOne, &sum));
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(300));
```
We need to use explicit synchronization here. Sleeping is a recipe for spurious test failures. We'll want to use something like a `condition_variable` to notify the main test thread that the work has finished.

chwarr replied:
Here's a quick and dirty event-like synchronization primitive:

```
class event
{
public:
    explicit event(bool initialValue = false) :
        _m(),
        _cv(),
        _set(initialValue)
    {}

    void set()
    {
        setResetImpl(true);
    }

    void reset()
    {
        setResetImpl(false);
    }

    void wait()
    {
        std::unique_lock<std::mutex> lock(_m);
        return _cv.wait(lock, [this]() {return _set;});
    }

    template <typename Rep, typename Period>
    bool wait(std::chrono::duration<Rep, Period> timeout)
    {
        std::unique_lock<std::mutex> lock(_m);
        return _cv.wait_for(lock, timeout, [this]() {return _set;});
    }

private:
    std::mutex _m;
    std::condition_variable _cv;
    bool _set;

    event(const event&) = delete;
    event(event&&) = delete;
    event& operator=(const event&) = delete;
    event& operator=(event&&) = delete;

    void setResetImpl(bool value)
    {
        {
            std::lock_guard<std::mutex> lock(_m);
            _set = value;
        }

        _cv.notify_all();
    }
};
```

chwarr commented on cpp/test/grpc/thread_pool.cpp:
```diff
@@ -0,0 +1,93 @@
+    static
+    void FinishAllTasksAfterDelete()
+    {
+        bond::ext::thread_pool* threads = new bond::ext::thread_pool(2);
```
Minor: `std::unique_ptr<thread_pool> threads(new bond::ext::thread_pool(2))` is preferred.

And, then, instead of `delete threads`, call `threads.reset()`.

chwarr commented on cpp/test/grpc/thread_pool.cpp:
```diff
@@ -0,0 +1,93 @@
+    void FinishAllTasksAfterDelete()
+    {
+        bond::ext::thread_pool* threads = new bond::ext::thread_pool(2);
+        int sum = 0;
```
`std::atomic<int> sum(0)` will allow us to get rid of the mutex.

chwarr commented on cpp/test/grpc/thread_pool.cpp:
```diff
@@ -0,0 +1,93 @@
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+            std::lock_guard<std::mutex> lock(sum_mutex);
+            sum++;
+        });
```
Consider the following instead:

```
auto increment = [&sum, &sum_mutex](){
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
            std::lock_guard<std::mutex> lock(sum_mutex);
            sum++;
        };

threads->schedule(increment);
threads->schedule(increment);
threads->schedule(increment);
threads->schedule(increment);
```

chwarr commented on examples/cpp/grpc-helloworld/helloworld.cpp:
```diff
@@ -0,0 +1,114 @@
+{
+    bond::ext::thread_pool threads(2);
+
+    std::string server_address("0.0.0.0:50051");
```
We should listen either on `127.0.0.1:50051` (or `[::1]:50051`) not `0.0.0.0` (which requires causes a firewall prompt on Windows).

chwarr commented on examples/cpp/grpc-helloworld/helloworld.cpp:
```diff
@@ -0,0 +1,114 @@
+        //wait for server to be alive
+        std::this_thread::sleep_for(std::chrono::milliseconds(200));
+
+        GreeterClient greeter(grpc::CreateChannel("localhost:50051", grpc::InsecureChannelCredentials()));
```
We need to connect to the same address that we listened on (i.e., use the same string). For example, on my machine `localhost` resolves to ::1 then 127.0.0.1.

chwarr commented on examples/cpp/grpc-helloworld/helloworld.cpp:
```diff
@@ -0,0 +1,114 @@
+
+    threads.schedule([&reply](){
+        //wait for server to be alive
+        std::this_thread::sleep_for(std::chrono::milliseconds(200));
```
Is there a deterministic way to wait for the server to be alive?
(Do we need to do this at all?)

chwarr commented on examples/cpp/grpc-helloworld/helloworld.cpp:
```diff
@@ -0,0 +1,114 @@
+        reply = greeter.SayHello(user);
+    });
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(300));
```
We need to wait for the reply. Possibly with a timeout to fail the test sooner than an hour, but we need to wait.

chwarr replied:
I think we can do all of this without using the thread pool. Just make all the client requests in `main()`. The server created by the builder has its own thread pool to handle dispatching requests.

chwarr commented on examples/cpp/grpc-helloworld/helloworld.cpp:
```diff
@@ -0,0 +1,114 @@
+    std::this_thread::sleep_for(std::chrono::milliseconds(300));
+
+    threads.schedule([&server_ptr](){
+        auto deadline = std::chrono::system_clock::now() + std::chrono::microseconds(1000);
```
Can we use `std::chrono::steady_clock` or `GPR_CLOCK_MONOTONIC` instead so we don't have to worry about the system clock going backward?

(Do we even need this shutdown?)

chwarr replied:
This time is in microseconds, but the other delays are in milliseconds. Intentional?

chwarr commented:
When you address the comments, can you do it in fixup commits so I can diff what I just reviewed with the changes?

lalo committed:
fixup

chwarr commented on cpp/inc/bond/ext/grpc/bond_utils.h:
```diff
@@ -39,15 +38,13 @@ class SerializationTraits<bond::comm::message<T>, typename std::enable_if<bond::
     grpc_slice_unref(slice);
 
-    return Status(StatusCode::OK, "ok");
+    return Status(StatusCode::OK, "");
```
`return Status::OK;` if not that, then just `return Status(StatusCode::OK)`.

chwarr committed:
[grpc++] Avoid one buffer copy when deserializing

Before, we were making two copies of the buffer given to us by gRPC in
the implementation of grpc::SerializationTraits::Deserialize for Bond
messages. This has been replaced with use of grpc_byte_buffer_reader to
only make one copy: from the grpc_byte_buffer to
boost::shared_ptr<char[]> that we then us to create the bonded<T>.

chwarr commented on cpp/inc/bond/ext/grpc/bond_utils.h:
```diff
@@ -63,22 +60,21 @@ class SerializationTraits<bond::comm::message<T>, typename std::enable_if<bond::
     boost::shared_ptr<char []> buff = boost::make_shared_noinit<char []>(GRPC_SLICE_LENGTH(slice));
     std::memcpy(buff.get(), GRPC_SLICE_START_PTR(slice), GRPC_SLICE_LENGTH(slice));
```
I pushed a commit that gets rid of the copy that `grpc_byte_buffer_reader_readall` was making.

chwarr commented on cpp/inc/bond/ext/grpc/bond_utils.h:
```diff
@@ -0,0 +1,94 @@
+
+    grpc_slice_unref(slice);
+
+    return Status(StatusCode::OK, "");
```
Minor: `Status::OK` or `Status(StatusCode::OK)`. There's no need for the `""`.

chwarr commented on cpp/inc/bond/ext/grpc/thread_pool.h:
```diff
@@ -0,0 +1,98 @@
+#pragma once
```
Minor: copyright header

chwarr commented on cpp/test/grpc/main.cpp:
```diff
@@ -0,0 +1,37 @@
+#define BOOST_TEST_NO_MAIN
```
Minor: copyright header

chwarr commented on cpp/test/grpc/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+# To add a new unit test suite, call add_unit_test at the bottom of this file on
```
Minor: copyright header

lalo replied:
cmakelists usually dont have it so i skipped this one

chwarr commented on cpp/test/grpc/test_utils_grpc.h:
```diff
@@ -0,0 +1,57 @@
+#include <chrono>
```
Minor: copyright header

chwarr commented on cpp/test/grpc/thread_pool.cpp:
```diff
@@ -0,0 +1,80 @@
+
```
Minor: copyright header

chwarr commented on cpp/test/grpc/test_utils_grpc.h:
```diff
@@ -0,0 +1,57 @@
+#include <condition_variable>
+#include <mutex>
+
+class Event
```
Minor: prevailing style in Bond C++ is lower-case class names. (Prevailing, not universal. 😄)

chwarr commented on examples/cpp/grpc-helloworld/helloworld.cpp:
```diff
@@ -0,0 +1,91 @@
+
+int main()
+{
+    std::string server_address("127.0.0.1:50051");
```
Minor: can be `const`

chwarr commented on cpp/test/grpc/thread_pool.cpp:
```diff
@@ -0,0 +1,80 @@
+
+        threads.schedule(std::bind(f_addOne, &sum, &sum_event));
+
+        sum_event.wait(std::chrono::seconds(30));
```
If `sum_event.wait(...)` returns false, we should fail the test. Something like

```
bool waitResult = sum_event.wait(std::chrono::seconds(30));
UT_AssertIsTrue(waitResult); // test timed out -- if we can pass a message, pass that as the message
```

chwarr reviewed (changes requested):
We're pretty close.

lalo committed:
minor fixup

lalo committed:
fixup

chwarr commented:
Merged into the branch https://github.com/chwarr/bond/tree/grpc%2B%2B
